x/delayedack/keeper/invariants.go

```diff
@@ -6,108 +6,102 @@ import (
 
-			latestFinalizedStateIndex, found := k.rollappKeeper.GetLatestFinalizedStateIndex(ctx, rollapp.RollappId)
-			if !found {
-				continue
-			}
-			latestFinalizedStateInfo, found := k.rollappKeeper.GetStateInfo(ctx, rollapp.RollappId, latestFinalizedStateIndex.Index)
-			if !found {
-				continue
+// packetsCorrespondsToStatusHeight checks that all rollapp packets stored are set to adequate status for all heights up to the latest height
+func (k Keeper) packetsCorrespondsToStatusHeight(checkPackets checkPacketsFn, checkRollappFrozen bool) sdk.Invariant {
+	return func(ctx sdk.Context) (msg string, stop bool) {
+		for _, rollapp := range k.rollappKeeper.GetAllRollapps(ctx) {
+			msg = k.checkRollapp(ctx, rollapp, checkPackets, checkRollappFrozen)
+			if stop = msg != ""; stop {
+				break
 			}
-			rollappsFinalizedHeight[rollapp.RollappId] = latestFinalizedStateInfo.GetLatestHeight()
-
 		}
 
-		packets := k.GetAllRollappPackets(ctx)
-
-		for _, packet := range packets {
-			latestFinalizedHeight := rollappsFinalizedHeight[packet.RollappId]
-
-			if packet.ProofHeight > latestFinalizedHeight && packet.Status == commontypes.Status_FINALIZED {
-				msg += fmt.Sprintf("rollapp packet for the height should not be in finalized status. height=%d, rollapp=%s, status=%s\n", packet.ProofHeight, packet.RollappId, packet.Status)
-				return msg, true
-			}
-			if packet.ProofHeight <= latestFinalizedHeight && packet.Status != commontypes.Status_FINALIZED {
-				msg += fmt.Sprintf("rollapp packet for the height should be in finalized status. height=%d, rollapp=%s, status=%s\n", packet.ProofHeight, packet.RollappId, packet.Status)
-				return msg, true
-			}
-		}
-		return msg, false
+		return
 	}
 }
 
-// PacketsFromRevertedHeightsAreReverted checks that all rollapp packets stored for a rollapp reverted height are also reverted
-func PacketsFromRevertedHeightsAreReverted(k Keeper) sdk.Invariant {
-	return func(ctx sdk.Context) (string, bool) {
-		var msg string
+func (k Keeper) checkRollapp(ctx sdk.Context, rollapp rtypes.Rollapp, checkPackets checkPacketsFn, checkFrozen bool) (msg string) {
+	if checkFrozen && !rollapp.Frozen {
+		return
+	}
+
+	// will stay 0 if no state is found
+	// but will still check packets
+	var latestFinalizedHeight uint64
 
-		rollapps := k.rollappKeeper.GetAllRollapps(ctx)
+	defer func() {
+		packets := k.ListRollappPackets(ctx, ByRollappID(rollapp.RollappId))
+		msg = checkPackets(packets, latestFinalizedHeight)
+	}()
 
-		frozenRollappsFinalizedHeight := make(map[string]uint64)
+	latestFinalizedStateIndex, found := k.rollappKeeper.GetLatestFinalizedStateIndex(ctx, rollapp.RollappId)
+	if !found {
+		return
+	}
 
-		for _, rollapp := range rollapps {
-			if !rollapp.Frozen {
-				continue
-			}
-			frozenRollappsFinalizedHeight[rollapp.RollappId] = 0
+	latestFinalizedStateInfo, found := k.rollappKeeper.GetStateInfo(ctx, rollapp.RollappId, latestFinalizedStateIndex.Index)
```

Comment:
Id say this is invariant breaking (i.e index without matching stateInfo) so maybe worth wrapping in `Must`.

Reply:
agreed, I tried my best to stick with the original logic